**Tasks**

Q1.

In [2]:
import pandas as pd
import numpy as np

from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score, confusion_matrix
df = pd.read_csv("Dataset 2.csv")

print("First Five Records:")
print(df.head())

First Five Records:
   UserID  Age  Gender SubscriptionType  WatchHoursPerWeek  DevicesUsed  \
0    1001   22  Female            Basic                 23            5   
1    1002   55    Male            Basic                  9            4   
2    1003   49    Male            Basic                  8            3   
3    1004   39  Female          Premium                 19            5   
4    1005   38  Female          Premium                 23            5   

  FavoriteGenre  AdClicks  MonthlySpend SubscriptionRenewed  
0        Comedy        13           353                  No  
1         Drama        14           317                 Yes  
2        Comedy        16           309                  No  
3         Drama        45           833                 Yes  
4        Sci-Fi        24           804                 Yes  


Q2.

In [3]:
print("\nDataset Shape:")
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])


Dataset Shape:
Rows: 750
Columns: 10


Q3.

In [4]:
print("\nColumn Names:")
print(df.columns.tolist())


Column Names:
['UserID', 'Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek', 'DevicesUsed', 'FavoriteGenre', 'AdClicks', 'MonthlySpend', 'SubscriptionRenewed']


Q4.

In [5]:
print("\nNumerical Features:")
print(df.select_dtypes(include=np.number).columns.tolist())

print("\nCategorical Features:")
print(df.select_dtypes(include='object').columns.tolist())


Numerical Features:
['UserID', 'Age', 'WatchHoursPerWeek', 'DevicesUsed', 'AdClicks', 'MonthlySpend']

Categorical Features:
['Gender', 'SubscriptionType', 'FavoriteGenre', 'SubscriptionRenewed']


Q5.

In [6]:
print("\nMissing Values:")
print(df.isnull().sum())


Missing Values:
UserID                 0
Age                    0
Gender                 0
SubscriptionType       0
WatchHoursPerWeek      0
DevicesUsed            0
FavoriteGenre          0
AdClicks               0
MonthlySpend           0
SubscriptionRenewed    0
dtype: int64


Q6.

In [7]:
avg_age = df["Age"].mean()
print("\nAverage Age:", round(avg_age, 2))


Average Age: 41.82


Q7.

In [8]:
avg_watch = df["WatchHoursPerWeek"].mean()
print("Average Watch Hours:", round(avg_watch, 2))

Average Watch Hours: 14.24


Q8.

In [9]:
avg_spend = df["MonthlySpend"].mean()
print("Average Monthly Spending:", round(avg_spend, 2))

Average Monthly Spending: 689.91


Q9.

In [10]:
print("\nSubscription Type Counts:")
print(df["SubscriptionType"].value_counts())


Subscription Type Counts:
SubscriptionType
Basic      342
Premium    279
VIP        129
Name: count, dtype: int64


Q10.

In [11]:
renewal_percentage = (
    (df["SubscriptionRenewed"] == "Yes").mean()
) * 100

print("\nRenewal Percentage:",
      round(renewal_percentage, 2), "%")


Renewal Percentage: 46.27 %


Q11.

In [12]:
df_encoded = df.copy()

label_encoders = {}

categorical_cols = [
    "Gender",
    "SubscriptionType",
    "FavoriteGenre",
    "SubscriptionRenewed"
]

for col in categorical_cols:
    le = LabelEncoder()
    df_encoded[col] = le.fit_transform(df_encoded[col])
    label_encoders[col] = le

print("\nEncoded Dataset:")
print(df_encoded.head())


Encoded Dataset:
   UserID  Age  Gender  SubscriptionType  WatchHoursPerWeek  DevicesUsed  \
0    1001   22       0                 0                 23            5   
1    1002   55       1                 0                  9            4   
2    1003   49       1                 0                  8            3   
3    1004   39       0                 1                 19            5   
4    1005   38       0                 1                 23            5   

   FavoriteGenre  AdClicks  MonthlySpend  SubscriptionRenewed  
0              1        13           353                    0  
1              2        14           317                    1  
2              1        16           309                    0  
3              2        45           833                    1  
4              5        24           804                    1  


Q12.

In [13]:
X = df_encoded.drop("SubscriptionRenewed", axis=1)
y = df_encoded["SubscriptionRenewed"]

print("\nFeature Matrix Shape:", X.shape)
print("Target Shape:", y.shape)


Feature Matrix Shape: (750, 9)
Target Shape: (750,)


Q13.

In [14]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42
)

print("\nTraining Samples:", len(X_train))
print("Testing Samples:", len(X_test))


Training Samples: 600
Testing Samples: 150


Q14.

In [15]:
dt_model = DecisionTreeClassifier(
    random_state=42
)

dt_model.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

Q15.

In [16]:
dt_pred = dt_model.predict(X_test)

dt_accuracy = accuracy_score(
    y_test,
    dt_pred
)

print("\nDecision Tree Accuracy:",
      round(dt_accuracy * 100, 2), "%")


Decision Tree Accuracy: 56.67 %


Q16.

In [17]:
cm = confusion_matrix(
    y_test,
    dt_pred
)

print("\nConfusion Matrix:")
print(cm)


Confusion Matrix:
[[50 32]
 [33 35]]


Q17.

In [18]:
knn_model = KNeighborsClassifier(
    n_neighbors=5
)

knn_model.fit(X_train, y_train)

KNeighborsClassifier()

Q18.

In [19]:
knn_pred = knn_model.predict(X_test)

knn_accuracy = accuracy_score(
    y_test,
    knn_pred
)

print("\nKNN Accuracy:",
      round(knn_accuracy * 100, 2), "%")

print("\nModel Comparison")
print("------------------")
print("Decision Tree:", round(dt_accuracy*100,2), "%")
print("KNN          :", round(knn_accuracy*100,2), "%")

if knn_accuracy > dt_accuracy:
    print("KNN performed better.")
else:
    print("Decision Tree performed better.")


KNN Accuracy: 60.0 %

Model Comparison
------------------
Decision Tree: 56.67 %
KNN          : 60.0 %
KNN performed better.


Q19.

In [20]:
df_reg = df_encoded.copy()

X_reg = df_reg.drop("MonthlySpend", axis=1)
y_reg = df_reg["MonthlySpend"]

X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
    X_reg,
    y_reg,
    test_size=0.20,
    random_state=42
)

lr_model = LinearRegression()

lr_model.fit(
    X_train_reg,
    y_train_reg
)

LinearRegression()

Q20.

In [21]:
sample_user = pd.DataFrame({
    "UserID":[9999],
    "Age":[30],
    "Gender":[1],
    "SubscriptionType":[1],
    "WatchHoursPerWeek":[15],
    "DevicesUsed":[2],
    "FavoriteGenre":[1],
    "AdClicks":[20],
    "SubscriptionRenewed":[1]
})

predicted_spend = lr_model.predict(sample_user)

print("\nPredicted Monthly Spending:")
print("₹", round(predicted_spend[0], 2))


Predicted Monthly Spending:
₹ 703.45


**Business Reflection Questions**

---

**1**. **Which factors appear to influence subscription renewal the most?**

Using the Decision Tree model's feature importance scores, the strongest predictors of subscription renewal were:

* MonthlySpend
* Age
* WatchHoursPerWeek
* AdClicks
* DevicesUsed / FavoriteGenre (moderate influence)

---

**2. Why is subscription renewal a classification problem?**

Subscription renewal is a classification problem because the target variable has a limited set of categories:

* Yes = User renewed
* No = User did not renew

The model's task is to assign each user to one of these discrete classes rather than predict a numerical value.

---

**3. Why is monthly spending a regression problem?**

Monthly spending is a regression problem because the target variable is a continuous numerical value.

*Examples:* ₹317, ₹804.

A regression model predicts an exact or approximate numeric amount rather than choosing between categories.

---

**4. Which algorithm performed better for renewal prediction?**

*Algorithm	Accuracy*

Decision Tree:	56.67%, K-Nearest Neighbors (K=5):	60.00%

KNN performed better than the Decision Tree model, achieving approximately 60% accuracy compared with 56.67% for the Decision Tree.

---

**5. How could the platform use these predictions to improve customer retention?**

Netflix could use renewal predictions in several ways:

a) *Identify At-Risk Users*

Users predicted not to renew could receive: Discount offers, Free trial extensions, Personalized recommendations

b) *Improve User Engagement*

Since watch hours and spending are important factors:
* Recommend content based on viewing history.
* Encourage usage through notifications and personalized suggestions.

c) *Optimize Subscription Plans*

Analyze which subscription types have lower renewal rates and:
* Adjust pricing.
* Add features to increase value.

d) *Targeted Marketing Campaigns*

Focus retention campaigns on users with:
* Low watch hours
* Low spending
* Declining engagement

e) *Revenue Forecasting*

Predicting renewals helps estimate future subscriber counts and revenue, supporting business planning and budgeting.